File: jas_ocaml/bin/svg_roundtrip.ml
(** CLI tool for cross-language commutativity testing.

    Usage:
      svg_roundtrip parse <file.svg>      -- parse SVG, output canonical JSON
      svg_roundtrip roundtrip <file.svg>  -- parse SVG, re-serialize, output SVG *)

let () =
  if Array.length Sys.argv < 3 then begin
    Printf.eprintf "Usage: %s parse|roundtrip <file.svg>\n" Sys.argv.(0);
    exit 1
  end;
  let mode = Sys.argv.(1) in
  let file = Sys.argv.(2) in
  let ic = open_in file in
  let n = in_channel_length ic in
  let svg = Bytes.create n in
  really_input ic svg 0 n;
  close_in ic;
  let doc = Jas.Svg.svg_to_document (Bytes.to_string svg) in
  match mode with
  | "parse" ->
    print_string (Jas.Test_json.document_to_test_json doc)
  | "roundtrip" ->
    print_string (Jas.Svg.document_to_svg doc)
  | _ ->
    Printf.eprintf "Unknown mode: %s (use 'parse' or 'roundtrip')\n" mode;
    exit 1

File: JasSwift/Tools/SvgRoundtrip.swift
/// CLI tool for cross-language commutativity testing.
///
/// Usage:
///   SvgRoundtrip parse <file.svg>      -- parse SVG, output canonical JSON
///   SvgRoundtrip roundtrip <file.svg>  -- parse SVG, re-serialize, output SVG

import Foundation
import JasLib

let args = CommandLine.arguments
guard args.count >= 3 else {
    fputs("Usage: \(args[0]) parse|roundtrip <file.svg>\n", stderr)
    exit(1)
}

let mode = args[1]
let path = args[2]

guard let data = FileManager.default.contents(atPath: path),
      let svg = String(data: data, encoding: .utf8) else {
    fputs("Failed to read: \(path)\n", stderr)
    exit(1)
}

let doc = svgToDocument(svg)

switch mode {
case "parse":
    print(documentToTestJson(doc), terminator: "")
case "roundtrip":
    print(documentToSvg(doc), terminator: "")
default:
    fputs("Unknown mode: \(mode) (use 'parse' or 'roundtrip')\n", stderr)
    exit(1)
}
